Add support for earlier macOS releases

cfelder opened the pull request:
Using macos_version_min_flags the lowest supported
version can be configured in the build scripts.
With his commit this is 10.15 for amd64 and 11.0 for arm64.

This has been tested w/ a built on macOS Monterey (12.6) which
runs on 12.6 itself as well as on Catalina (10.15).

cfelder committed:
Add support for earlier macOS releases

Using macos_version_min_flags the lowest supported
version can be configured in the build scripts.
With his commit this is 10.15 for amd64 and 11.0 for arm64.

Fixes: containers/podman#16048
Signed-off-by: Christian Felder <c.felder@fz-juelich.de>

anjannath commented:
@cfelder few questions (don't know c/c++) does this add a constraint to the build host's macOS version now or we can build it on any version and minimum supported will still be `10.15`? 

and if this flag is not used then the target is the current macOS version as the minimum supported one?

cfelder committed:
Add MACOSX_DEPLOYMENT_TARGET setting

Additionally to setting the `macos_version_min_flags` the
MACOS_DEPLOYMENT_TARGET must be set to create the corresponding entries
for the LC_BUILD_VERSION load command.

The load commands can be listed using `otool`:

    $ otool -l {binary}

          cmd LC_BUILD_VERSION
      cmdsize 32
     platform 1
        minos 10.15
          sdk 12.3
       ntools 1
         tool 3
      version 819.6
    Load command 11

The resulting binary has been created using the XCode SDK 12.3 and should run
on macOS version 10.15 and later.

For a more detailled explanation, I am citing the XCode developer docs below:

> MACOSX_DEPLOYMENT_TARGET
>
> Code will load on this and later versions of macOS. Framework APIs that are
> unavailable in earlier versions will be weak-linked; your code should check
> for null function pointers or specific system versions before calling newer
> APIs.

Signed-off-by: Christian Felder <c.felder@fz-juelich.de>

cfelder commented:
> @cfelder few questions (don't know c/c++) does this add a constraint to the build host's macOS version now or we can build it on any version and minimum supported will still be `10.15`?

The build host must have the XCode SDK 10.15 or newer. The resulting binaries should run on macOS 10.15 and newer. I've tested this w/ building and running on my macOS Monterey 12.6 machine and running the same binaries on my macOS Catalina 10.15.7 machine.

> and if this flag is not used then the target is the current macOS version as the minimum supported one?

Yes.

See also a more detailed explanation about MACOSX_DEPLOYMENT_TARGET from the Xcode Developer docs:

> MACOSX_DEPLOYMENT_TARGET
> Code will load on this and later versions of macOS. Framework APIs that are unavailable in earlier versions will be weak-linked; your code should check for null function pointers or specific system versions before calling newer APIs.
